stl/inc/coroutine

```diff
@@ -0,0 +1,337 @@
+// * `noop_coroutine_handle` makes no sense after a slicing assignment. Maybe it should have a conversion to
+//   `coroutine_handle<>` rather than being publicly derived? (This interacts with the "hidden friend" comparison
+//   operator question.)
+// * [coroutine.trivial.awaitables]/1 needs a cross-reference for *await-expression*.
+//
+// * [coroutine.handle]
+//   * Replace para 1:
+//     An object whose type is a specialization of `coroutine_handle` is called a *coroutine handle* and can be used to
+//     refer to a suspended or executing coroutine. A coroutine handle refers to a coroutine if and only if the return
+//     value of `address()` is non-null.
+//   * Replace para 2:
+//     A program shall not declare an explicit or partial specialization of `coroutine_handle`.
+// * coroutine_handle<Promise>::from_promise:
+//   Expects: `p` references the promise object of a coroutine.
+//   Returns: A coroutine handle `h` that refers to the coroutine whose promise object is referenced by `p`. [Note:
+//            `addressof(h.promise()) == addressof(p)` is consequently `true`.-end note]
+//   Throws: Nothing.
+// * coroutine_handle<>::address:
+//   Returns: A non-null pointer value if and only if `*this` refers to a coroutine.
+// * coroutine_handle<>::from_address and coroutine_handle<Promise>::from_address:
+//   Expects: ...
+//   Returns: A coroutine handle that refers to the same coroutine, if any, as the coroutine handle from which `addr`
+//            was obtained. [Note: For a coroutine handle `h`, `from_address(h.address()) == h` is `true`.]
+// * coroutine_handle<>::done: s/Requires/Expects/
+// * coroutine_handle<>::operator() and coroutine_handle<>::resume:
+//   Expects: `*this` refers to a coroutine suspended at other than its final suspend point.
+//   Effects: Resumes the execution of the coroutine.
+// * coroutine_handle<>::destroy: s/Requires/Expects/
+// * coroutine_handle<Promise>::promise: s/Requires/Expects/
+
+// intrinsics used in the implementation of coroutine_handle
+_EXTERN_C
+size_t _coro_resume(void*); // FIXME: Calling convention?
+#pragma intrinsic(_coro_resume) // FIXME: Is `#pragma intrinsic` necessary?
+
+void _coro_destroy(void*); // FIXME: Calling convention?
+#pragma intrinsic(_coro_destroy) // FIXME: Is `#pragma intrinsic` necessary?
+_END_EXTERN_C
+
+_STD_BEGIN
+
+// STRUCT TEMPLATE coroutine_traits
+template <class _Ret, class = void>
+struct _Coroutine_traits {};
+
+template <class _Ret>
+struct _Coroutine_traits<_Ret, void_t<typename _Ret::promise_type>> {
+    using promise_type = typename _Ret::promise_type;
+};
+
+template <class _Ret, class...>
+struct coroutine_traits : _Coroutine_traits<_Ret> {};
+
+// STRUCT TEMPLATE coroutine_handle
+template <class = void>
+struct coroutine_handle;
+
+struct _Resumable_frame {
+    using _Fn = void __cdecl(void*);
+
+    _Fn* _Resume;
+    unsigned short _Index;
+    unsigned short _Flags;
+};
+
+inline constexpr intptr_t _Noop_sentinel_value = -1;

```

Comment:
You are never using the value directly, but always in conjunction with a 'reinterpret_cast' would it be possible to directly define
```cpp
inline constexpr void* _Noop_sentinel_pointer = reinterpret_cast<void*>(intptr_t{-1});
```

Reply:
Derp, `reinterpret_cast` is not constexpr. Given that this is essentially a fixed bit-pattern why not `static_cast`

Reply:
static_cast cannot convert integer types to pointer types

Reply:
Yep!